tests/php/test-amp-image-dimension-extractor.php

```diff
@@ -184,4 +184,61 @@ public function test__amp_wp_user_agent() {
 
 		$this->assertEquals( $expected, $user_agent );
 	}
+
+	/**
+	 * @covers \AMP_Image_Dimension_Extractor::extract_by_filename_or_filesystem
+	 */
+	public function test_extract_by_filename_or_filesystem() {
+
+		$attachment_id = $this->factory()->attachment->create_upload_object( __DIR__ . '/data/images/wordpress-logo.png' );
+
+		$full_image       = wp_get_attachment_image_src( $attachment_id, 'full' );
+		$thumbnail_image  = wp_get_attachment_image_src( $attachment_id, 'thumbnail' );
+		$external_image   = 'https://example.com/wp-content/uploads/2021/04/American_bison_k5680-1-1024x668.jpg';
+		$external_image_1 = 'https://via.placeholder.com/1500/000.png/FF0';
+		$external_image_2 = 'https://via.placeholder.com/1000/000.png/FF0';
+
+		$expected = [
+			$full_image[0]      => [
+				'width'  => $full_image[1],
+				'height' => $full_image[2],
+			],
+			$thumbnail_image[0] => [
+				'width'  => $thumbnail_image[1],
+				'height' => $thumbnail_image[2],
+			],
+			$external_image     => [
+				'width'  => 1024,
+				'height' => 668,
+			],
+			$external_image_1   => [],
+			$external_image_2   => [
+				'width'  => 1000,
+				'height' => 1000,
+			],
+		];
+
+		$input = [
+			$full_image[0]      => [],
+			$thumbnail_image[0] => [],
+			$external_image     => [],
+			$external_image_1   => [],
+			$external_image_2   => [
+				'width'  => 1000,
+				'height' => 1000,
+			],
+		];
```

Comment:
To facilitate maintenance, I suggest that these two be combined into one array, like:

```php
$data = [
    $full_image[0] => [
        'input' => [],
        'expected' => [
            'width'  => $full_image[1],
            'height' => $full_image[2],
        ],
    ],
    ...
]
```

Then you can extract the `$expected` and `$input` via `array_map()`.